libsplinter/src/template/parser.rs

```diff
@@ -0,0 +1,282 @@
+
+#[derive(Deserialize, Debug)]
+struct CircuitCreateTemplate {
+    args: Vec<YamlRuleArgument>,
+    rules: HashMap<String, serde_yaml::Value>,
+}
+
+#[derive(Deserialize, Debug)]
+struct YamlRuleArgument {
+    name: String,
+    required: bool,
+    #[serde(rename = "default")]
+    default_value: Option<String>,
+}
+
+pub struct YamlCreateCircuitTemplateParser {
+    circuit_rules: CreateCircuitRules,
+    service_rules: ServiceDefinitionRules,
+    rule_arguments: HashMap<String, Vec<RuleArgument>>,
+}
+
+impl Default for YamlCreateCircuitTemplateParser {
+    fn default() -> Self {
+        let mut rule_parser = YamlCreateCircuitTemplateParser {
+            circuit_rules: CreateCircuitRules::default(),
+            service_rules: ServiceDefinitionRules::default(),
+            rule_arguments: HashMap::new(),
+        };
+        let rule = CircuitManagementTypeRule::default();
+        rule_parser.register_circuit_rule(&rule.name(), Box::new(rule));
+        rule_parser
+    }
+}
+
+impl YamlCreateCircuitTemplateParser {
+    pub fn parse_rules(
+        mut self,
+        path: &str,
+        arg_values: HashMap<String, String>,
+    ) -> Result<CreateCircuitBuilder, TemplateParserError> {
+        let file = File::open(path)?;
+
+        let template: CircuitCreateTemplate = serde_yaml::from_reader(file)?;
+
+        for (rule_name, value) in template.rules.iter() {
+            self.apply_rule(rule_name, &arg_values, &serde_yaml::to_vec(value)?)?
+        }
+
+        let mut builder = self.circuit_rules.builder();
+
+        if self.service_rules.has_builders() {
+            let services = self.service_rules.build_services()?;
+            builder = builder.with_roster(&services);
+        }
+
+        Ok(builder)
+    }
+
+    pub fn register_service_rule(
+        &mut self,
+        name: &str,
+        rule: Box<dyn Rule<Vec<SplinterServiceBuilder>>>,
+    ) {
+        let arguments = rule.get_arguments();
+        self.rule_arguments.insert(name.to_string(), arguments);
+        self.service_rules.register_rule(name, rule);
+    }
+
+    pub fn register_circuit_rule(&mut self, name: &str, rule: Box<dyn Rule<CreateCircuitBuilder>>) {
+        let arguments = rule.get_arguments();
+        self.rule_arguments.insert(name.to_string(), arguments);
+        self.circuit_rules.register_rule(name, rule);
+    }
+
+    pub fn apply_rule(
+        &mut self,
+        name: &str,
+        args: &HashMap<String, String>,
+        values: &[u8],
+    ) -> Result<(), TemplateParserError> {
+        if self.circuit_rules.is_circuit_rule(name) {
+            self.circuit_rules.apply_rule(name, args, values)?
+        }
+
+        if self.service_rules.is_service_rule(name) {
+            self.service_rules.apply_rule(name, args, values)?
+        }
+        Ok(())
+    }
+
+    pub fn arguments(&self) -> HashMap<String, Vec<RuleArgument>> {
+        self.rule_arguments.clone()
+    }
+}
+
+struct CreateCircuitRules {
```

Comment:
I think we shouldn't have different types of rules (circuit vs service); any rule should operate on the entirety of the circuit (which includes defining or operating on the services within the circuit).